support the common data model

richm opened the pull request:
This adds 3 new elasticsearch.yml openshift.config section parameters:

* `use_common_data_model` - boolean - default `false` - if true, tell the
  plugin that you want to use the common data model
* `project_index_prefix` - string - default `""` - set this to the string value
  that project/namespace indices use as their prefix - for example, with the
  common data model, if the namespace is `test`, the index name will be
  `project.test.$uuid.YYYY.MM.DD`.  In this case, use `"project."` as the
  prefix.
* `time_field_name` - string - default `"time"` - the common data model uses
  `"@timestamp"` as the name of the time field

If the common data model is used, the common data model index-pattern values
will be loaded to match the common data model index naming, model, and field
naming conventions.

The index-mapping values are from doing a search of the .kibana index after
loading Kibana and loading data from common data model indices.

richm commented:
@ewolinetz @jcantrill PTAL

richm commented:
@ewolinetz In addition to these changes, are there any acl/search guard related changes needed?

ewolinetz commented:
@richm yes, it looks like the logic for that is missing from here -- and it may change the seeding logic slightly. The ACL is defined and updated within the dynamic_acl_filter

richm commented:
@ewolinetz are you talking about this code from `public void syncFrom`?
```
	public void syncFrom(UserProjectCache cache, final String userProfilePrefix){
...
			String indexName = String.format("%s?*", project.replace('.', '?'));
...
			String indexName = String.format("%s?%s", userProfilePrefix.replace('.', '?'), usernameHash);
```
Those index names need to have the `project.` prefix?

richm commented:
yeah, not sure what's going on.  I created a testuser that has admin rights in the test project.  If I login to the openshift console, I can only see the test project.  However, in kibana/es, I can view project.logging.* and even .operations.*

richm commented:
I went back to 1.3 so I could see how access control is supposed to work, but I'm having a hard time testing access control with 1.3.  This is what I did:
```
oc new-project test
oc new-app centos/ruby-22-centos7~https://github.com/openshift/ruby-ex.git
oc login --username=testuser --password=testuser
test_token="$(oc whoami -t)"
test_name="$(oc whoami)"
test_ip="127.0.0.1"
oc login --username=system:admin
oc policy add-role-to-user admin testuser -n test
oc exec -n logging logging-kibana-1-napr6 -c kibana -- curl -s -k --cert /etc/kibana/keys/cert --key /etc/kibana/keys/key -H "X-Proxy-Remote-User: $test_name" -H "Authorization: Bearer $test_token" -H "X-Forwarded-For: 127.0.0.1" https://logging-es:9200/test.*/_search | python -mjson.tool
{
    "error": "UnknownHostException[No trusted proxies]",
    "status": 500
}
```
but if I omit the proxy stuff, it works:
```
oc exec -n logging logging-kibana-1-napr6 -c kibana -- curl -s -k        --cert /etc/kibana/keys/cert --key /etc/kibana/keys/key        https://logging-es:9200/test.*/_search | python -mjson.tool
{
    "_shards": {
        "failed": 0,
        "successful": 5,
        "total": 5
    },
    "hits": {
        "hits": [
            {
...
            }
        ],
        "max_score": 1.0,
        "total": 45
```

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/DocumentBuilder.java:
```diff
@@ -54,14 +57,20 @@ public DocumentBuilder intervalName(String name) {
-	public DocumentBuilder applicationFields() {
-		return addField("fields", StringEscapeUtils.escapeJava(APPLICATION_FIELDS));
+	public DocumentBuilder applicationFields(boolean use_cdm) {
+		if ( false == use_cdm )
```
This is an odd coding convention for java though I like the explicitness of it.  Consider using !use_cdm but its not a deal breaker for me.

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/DocumentBuilder.java:
```diff
@@ -54,14 +57,20 @@ public DocumentBuilder intervalName(String name) {
+		if ( false == use_cdm )
+			return addField("fields", StringEscapeUtils.escapeJava(APPLICATION_FIELDS));
+		else
+			return addField("fields", StringEscapeUtils.escapeJava(APPLICATION_CDM_FIELDS));
```
We should pull this into a constant since its used in many places.

richm replied:
@jcantrill sorry - pull what into a constant?

jcantrill replied:
"fields"

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/kibana/KibanaSeed.java:
```diff
@@ -48,11 +49,12 @@
 import org.elasticsearch.transport.RemoteTransportException;
 
-public class KibanaSeed {
+public class KibanaSeed implements ConfigurationSettings {
```
Why do we need to impl ConfigurationSettings?  I didnt see any obvious uses of it.

richm replied:
@jcantrill that seems to be the technique that other classes use in order to have access to the constants defined there such as OPENSHIFT_CONFIG_USE_COMMON_DATA_MODEL

ewolinetz replied:
Correct

ewolinetz commented:
@richm the project we get in that snippet should already have the prefix at that point in time. That is just for formatting the ACL to match searchguard's required format (they don't allow for '.' so we replace them with '?').

If a normal user is able to see everything, then something is wrong with their ACL... if you change the logger level for the kibana seed class, you should see which projects come back for the user.

richm commented:
With this patch plus https://github.com/fabric8io/openshift-elasticsearch-plugin/pull/47 everything is working - testuser can see `project.test.*` but not `project.logging.*` or `.operations.*`:
```
{"error":{"root_cause":[{"type":"security_exception","reason":"no permissions for indices:data/read/search"}],"type":"security_exception","reason":"no permissions for indices:data/read/search"},"status":403}
```
I think I addressed all of the review comments.  Please review.

ewolinetz commented:
@richm I wouldn't mind doing a release containing just #47 prior to merging this in just in case (call me paranoid).

Otherwise LGTM

richm commented:
@jcantrill ptal

@ewolinetz ok - yes, let's release #47 asap

jcantrill commented on src/main/java/io/fabric8/elasticsearch/plugin/acl/SearchGuardRoles.java:
```diff
@@ -156,17 +156,23 @@ public void removeRole(Roles role) {
+
+			// If using common data model, allow access to both the $projname.$uuid.* indices and
+			// the project.$projname.$uuid.* indices for backwards compatability
+			if ( cdm_project_prefix.length() > 0 ) {
```
StringUtils.IsBlank .... or isEmpty... will handle the case of empty string

jcantrill commented:
Other then minor nit... LGTM

richm commented:
I found a couple of other minor issues.  @ewolinetz @jcantrill ptal

jcantrill commented:
LGTM

ewolinetz commented on README.md:
```diff
@@ -43,6 +43,24 @@ openshift.operations.project.names: ["default", "openshift", "openshift-infra"]
+* `project_index_prefix` - string - default `""` - set this to the string value
+  that project/namespace indices use as their prefix - for example, with the
+  common data model, if the namespace is `test`, the index name will be
+  `project.test.$uuid.YYYY.MM.DD`.  In this case, use `"project."` as the
```
technically shouldn't they use "project" as the prefix? if they use "project." then our indices would end up as "project..$project.$uuid.$date"

I think we would want to discourage having a user provide the delimiter value in the case they don't provide it it could cause some grief

richm replied:
Ok.  It makes the logic easier if the delimiter is included with the prefix - I can do things like
```
		return project_prefix + project + ".*";
```
as opposed to
```
    if (StringUtils.isEmtpy(project_prefix))
        return project + ".*";
    else
        return project_prefix + "." + project + ".*";
```

richm commented:
@ewolinetz @jcantrill fixed review comments - PTAL

ewolinetz commented:
LGTM

richm commented:
[merge]

richm commented:
[merge]

ewolinetz commented:
[merge]
